File: backend/app/api/conversations.py
"""مسارات المحادثات والرسائل.

**كل مسار هنا يعمل على محادثات صاحب الرمز وحده.** لا يستقبل أي مسار جهة ولا
مالكًا: يُشتقان من الرمز داخل `conversation_service`. محادثة موظف آخر — ولو
كان زميلًا في الجهة نفسها — تعود **404 بلا بيانات**.
"""

from fastapi import APIRouter, HTTPException, Query, status

from ..schemas import (
    ConversationCreateRequest,
    ConversationListResponse,
    ConversationOut,
    ConversationRenameRequest,
    MessageListResponse,
    MessageOut,
    PageMeta,
)
from ..services.conversation_service import (
    ConversationNotFoundError,
    create_conversation,
    delete_conversation,
    get_conversation,
    list_conversations,
    list_messages,
    rename_conversation,
)
from ..services.conversation_store import Conversation, Message, Page
from .dependencies import CurrentUser

router = APIRouter(prefix="/api/conversations", tags=["conversations"])

_NOT_FOUND = {"description": "المحادثة غير موجودة لصاحب الرمز"}
_UNAUTHORIZED = {"description": "رمز الدخول مفقود أو غير صالح"}


def _to_conversation_out(conversation: Conversation) -> ConversationOut:
    return ConversationOut(
        id=conversation.id,
        title=conversation.title,
        message_count=conversation.message_count,
        created_at=conversation.created_at,
        updated_at=conversation.updated_at,
    )


def _to_message_out(message: Message) -> MessageOut:
    return MessageOut(
        id=message.id,
        conversation_id=message.conversation_id,
        role=message.role,
        content=message.content,
        created_at=message.created_at,
    )


def _page_meta(page: Page) -> PageMeta:
    return PageMeta(total=page.total, limit=page.limit, offset=page.offset)


def _not_found(exc: Exception) -> HTTPException:
    return HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))


@router.post(
    "",
    response_model=ConversationOut,
    status_code=status.HTTP_201_CREATED,
    summary="إنشاء محادثة",
    responses={401: _UNAUTHORIZED},
)
def add_conversation(
    payload: ConversationCreateRequest, user: CurrentUser
) -> ConversationOut:
    """ينشئ محادثة فارغة يملكها صاحب الرمز.

    لا يلزم استدعاؤه قبل المحادثة: `POST /api/chat` بلا `conversation_id`
    ينشئ محادثة تلقائيًا بعنوان مشتق من أول رسالة.
    """
    return _to_conversation_out(create_conversation(actor=user, title=payload.title))


@router.get(
    "",
    response_model=ConversationListResponse,
    summary="قائمة محادثات الموظف",
    responses={401: _UNAUTHORIZED},
)
def read_conversations(
    user: CurrentUser,
    limit: int | None = Query(None, ge=1, description="عدد المحادثات في الصفحة"),
    offset: int | None = Query(None, ge=0, description="عدد المحادثات المتجاوَزة"),
) -> ConversationListResponse:
    """يعيد محادثات صاحب الرمز من الأحدث تعديلًا.

    محادثات الزملاء لا تظهر هنا إطلاقًا، ولو كان صاحب الرمز مسؤول الجهة.
    """
    page = list_conversations(actor=user, limit=limit, offset=offset)
    return ConversationListResponse(
        conversations=[_to_conversation_out(item) for item in page.items],
        page=_page_meta(page),
    )


@router.get(
    "/{conversation_id}",
    response_model=ConversationOut,
    summary="قراءة محادثة",
    responses={401: _UNAUTHORIZED, 404: _NOT_FOUND},
)
def read_conversation(conversation_id: int, user: CurrentUser) -> ConversationOut:
    """يعيد بيانات محادثة يملكها صاحب الرمز."""
    try:
        return _to_conversation_out(
            get_conversation(actor=user, conversation_id=conversation_id)
        )
    except ConversationNotFoundError as exc:
        raise _not_found(exc) from exc


@router.patch(
    "/{conversation_id}",
    response_model=ConversationOut,
    summary="إعادة تسمية محادثة",
    responses={401: _UNAUTHORIZED, 404: _NOT_FOUND},
)
def edit_conversation(
    conversation_id: int, payload: ConversationRenameRequest, user: CurrentUser
) -> ConversationOut:
    """يغيّر عنوان محادثة يملكها صاحب الرمز."""
    try:
        return _to_conversation_out(
            rename_conversation(
                actor=user, conversation_id=conversation_id, title=payload.title
            )
        )
    except ConversationNotFoundError as exc:
        raise _not_found(exc) from exc


@router.delete(
    "/{conversation_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="حذف محادثة",
    responses={401: _UNAUTHORIZED, 404: _NOT_FOUND},
)
def remove_conversation(conversation_id: int, user: CurrentUser) -> None:
    """يحذف محادثة يملكها صاحب الرمز **ورسائلها معها**.

    حذف فعلي لا تعطيل، بخلاف حسابات الموظفين: المحادثة محتوى يملكه صاحبه وله
    أن يزيله، ولا يرتبط بها تاريخ يخص غيره.
    """
    try:
        delete_conversation(actor=user, conversation_id=conversation_id)
    except ConversationNotFoundError as exc:
        raise _not_found(exc) from exc


@router.get(
    "/{conversation_id}/messages",
    response_model=MessageListResponse,
    summary="رسائل محادثة",
    responses={401: _UNAUTHORIZED, 404: _NOT_FOUND},
)
def read_messages(
    conversation_id: int,
    user: CurrentUser,
    limit: int | None = Query(None, ge=1, description="عدد الرسائل في الصفحة"),
    offset: int | None = Query(None, ge=0, description="عدد الرسائل المتجاوَزة"),
) -> MessageListResponse:
    """يعيد رسائل المحادثة **بالترتيب الزمني الصاعد** مع الترقيم.

    من الأقدم إلى الأحدث لأن المحادثة تُقرأ من أولها، و`offset=0` يعطي بدايتها.
    """
    try:
        page = list_messages(
            actor=user, conversation_id=conversation_id, limit=limit, offset=offset
        )
    except ConversationNotFoundError as exc:
        raise _not_found(exc) from exc

    return MessageListResponse(
        messages=[_to_message_out(item) for item in page.items],
        page=_page_meta(page),
    )

File: frontend/lib/conversations.ts
/**
 * المحادثات والرسائل — نداءات `/api/conversations` الموجودة فعلًا في
 * `backend/app/api/conversations.py`.
 *
 * لا يُرسل أي نداء هنا `organization_id` ولا `user_id`: الجهة والمالك
 * يُشتقان من رمز الدخول في السيرفر.
 */

import { apiRequest } from "@/lib/api";

/** يطابق `ConversationOut` في `backend/app/schemas/workspace.py`. */
export type Conversation = {
  id: number;
  title: string;
  message_count: number;
  created_at: string;
  updated_at: string;
};

/** يطابق `MessageOut`. */
export type Message = {
  id: number;
  conversation_id: number;
  role: "user" | "assistant";
  content: string;
  created_at: string;
};

export type PageMeta = { total: number; limit: number; offset: number };

type ConversationListResponse = {
  conversations: Conversation[];
  page: PageMeta;
};

type MessageListResponse = {
  messages: Message[];
  page: PageMeta;
};

export function listConversations(
  token: string,
  signal?: AbortSignal,
): Promise<ConversationListResponse> {
  return apiRequest<ConversationListResponse>("/api/conversations", {
    token,
    signal,
  });
}

export function createConversation(
  token: string,
  title?: string,
): Promise<Conversation> {
  return apiRequest<Conversation>("/api/conversations", {
    method: "POST",
    body: { title: title ?? null },
    token,
  });
}

export function renameConversation(
  token: string,
  id: number,
  title: string,
): Promise<Conversation> {
  return apiRequest<Conversation>(`/api/conversations/${id}`, {
    method: "PATCH",
    body: { title },
    token,
  });
}

/** يعيد 204 بلا جسم — لذلك `void`. */
export function deleteConversation(token: string, id: number): Promise<void> {
  return apiRequest<void>(`/api/conversations/${id}`, {
    method: "DELETE",
    token,
  });
}

export function listMessages(
  token: string,
  id: number,
  signal?: AbortSignal,
): Promise<MessageListResponse> {
  return apiRequest<MessageListResponse>(`/api/conversations/${id}/messages`, {
    token,
    signal,
  });
}

/** أقصى طول عنوان، مطابق لـ`MAX_TITLE_LENGTH` في الـBackend. */
export const MAX_TITLE_LENGTH = 300;

export function validateTitle(value: string): string | null {
  const cleaned = value.trim();
  if (!cleaned) return "عنوان المحادثة لا يمكن أن يكون فارغًا.";
  if (cleaned.length > MAX_TITLE_LENGTH) {
    return `العنوان أطول من ${MAX_TITLE_LENGTH} حرفًا.`;
  }
  return null;
}
